src/display/canvas.js

```diff
@@ -1565,6 +1449,117 @@ class CanvasGraphics {
     this.ctx.restore();
   }
 
+  composeSMask(ctx, smask, layerCtx, layerBox) {
+    const layerOffsetX = layerBox[0];
+    const layerOffsetY = layerBox[1];
+    const layerWidth = layerBox[2] - layerOffsetX;
+    const layerHeight = layerBox[3] - layerOffsetY;
+    if (layerWidth === 0 || layerHeight === 0) {
+      return;
+    }
+    this.genericComposeSMask(
+      smask.context,
+      layerCtx,
+      layerWidth,
+      layerHeight,
+      smask.subtype,
+      smask.backdrop,
+      smask.transferMap,
+      layerOffsetX,
+      layerOffsetY,
+      smask.offsetX,
+      smask.offsetY
+    );
+    ctx.save();
+    ctx.globalAlpha = 1;
+    ctx.globalCompositeOperation = "source-over";
+    ctx.setTransform(1, 0, 0, 1, 0, 0);
+    ctx.drawImage(layerCtx.canvas, 0, 0);
+    ctx.restore();
+  }
+
+  genericComposeSMask(
+    maskCtx,
+    layerCtx,
+    width,
+    height,
+    subtype,
+    backdrop,
+    transferMap,
+    layerOffsetX,
+    layerOffsetY,
+    maskOffsetX,
+    maskOffsetY
+  ) {
+    let maskCanvas = maskCtx.canvas;
+    let maskX = layerOffsetX - maskOffsetX;
+    let maskY = layerOffsetY - maskOffsetY;
+
+    if (backdrop) {
+      if (
+        maskX < 0 ||
+        maskY < 0 ||
+        maskX + width > maskCanvas.width ||
+        maskY + height > maskCanvas.height
+      ) {
+        const canvas = this.cachedCanvases.getCanvas(
+          "maskExtension",
+          width,
+          height
+        );
+        const ctx = canvas.context;
+        ctx.drawImage(maskCanvas, -maskX, -maskY);
+        if (backdrop.some(c => c !== 0)) {
```

Comment:
Rather than checking this here and below, would it make sense to instead do that in the worker?
With the following change
```diff
diff --git a/src/core/evaluator.js b/src/core/evaluator.js
index b608feed6..fefd88160 100644
--- a/src/core/evaluator.js
+++ b/src/core/evaluator.js
@@ -515,6 +515,10 @@ class PartialEvaluator {
       if (smask?.backdrop) {
         colorSpace ||= ColorSpace.singletons.rgb;
         smask.backdrop = colorSpace.getRgb(smask.backdrop, 0);
+
+        if (smask.backdrop.every(c => c === 0)) {
+          smask.backdrop = null;
+        }
       }

       operatorList.addOp(OPS.beginGroup, [groupOptions]);
```
this line could become
```suggestion
        if (backdrop) {
```

Reply:
There's already a check here:
https://github.com/mozilla/pdf.js/pull/18029/files#diff-76166cff75cae0ec9c490ae47aebbdadff02dbc1e26caaf93779a1bace64dc47R1498
and if we do what you're proposing, then we won't enter in the above block.
Or am I missing something ?

Reply:
Missed that, sorry!